OSRF migration: Adding `nodl_to_policy` tool

aprotyas opened the pull request:
This PR serves as a repository migration from [aprotyas/nodl_to_policy](github.com/aprotyas/nodl_to_policy) to [osrf/nodl_to_policy](github.com/osrf/nodl_to_policy).

The PR includes:
- README with build/use instructions.
- `ros2cli` entrypoint and a `convert` verb.
- NoDL &rarr; ROS 2 access control policy API definition.
- Unit tests for the `convert` verb.

Unit tests for the actual conversion module will be submitted as a separate PR.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Initial commit

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
`ament_python` package skeleton added

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Importing `ubuntu-robotics/nodl` dependencies

Also:

* Updating some package information in `setup.py`
* Adding a `.gitignore` from
  (https://github.com/github/gitignore/blob/master/Python.gitignore)

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Adding license badge to README

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Added a `ros2cli` entrypoint, and a `convert` verb

Usage: `ros2 nodl_to_policy convert [NoDL description file(s)]`

The API for the conversion mechanism has not been well-defined yet, but
will be part of the next steps.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Fixing `ament_pycodestyle` lint errors

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Adding policy XML positional argument to `convert`

The `convert` verb requires both positional arguments for the NoDL
description file(s) and the ROS 2 access control policy file. The latter
was not previously part of `argparse.Parser`

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Separate conversion API and the `convert` CLI verb

This commit gives better structure to the `_ConvertVerb` class.
Notably:

* The actual conversion API has been stripped away from this class.
* The verb extension parses provided NoDL descriptions, and makes
calls to policy conversion/write functions, as will be defined shortly.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Partially defining the policy conversion API

In this commit:
* Sketched some ideas, in comments, about the conversion process.
* Defined a simple `write_policy` that relies on the XSLT operation
performed in `sros2.policy.dump_policy` for successful policy XML
file generation.

Confusions:
* What to do about the enclave paths? I'm not sure if there's a
simple, non-intuitive way to extract node process paths from the NoDL
description.
* Does the `xpointer` and `xinclude` shenanigans in the example ROS 2
access control policy files get placed through the transpiling in
`sros2.policy.dump_policy`? If not, where do they come into play?

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Fixing module import errors in _ConvertVerb module

Also, reordering the arguments to `write_policy` to match
the API declared in the `nodl_to_policy.policy` module

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
"Complete"-ed NoDL conversion API

Concerns:
* Enclave paths are not well specified. Is it fair to make an
assumption and default to `"[node_name][executable_name]"`?
* Node namespaces are also not well specified. It certainly is
not fair to assume all nodes belong to the unspecified `"/"`
general namespace.

Both of the above concerns can be addressed by expanding the
definition of the node interface description language. This
discussion probably has a place in ros2/design#266.

TO-DO:
* Verify whether default enclave paths should be left blank
or should be set to an unqualified general path ("/")
* Writing unit and functional tests for the `nodl_to_policy`
utility.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

aprotyas committed:
Initial development of the NoDL conversion module

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Added `--print` argument to display the policy XML

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

aprotyas committed:
Clarify utility purpose on README

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Added `mypy` and `pycodestyle` unit tests

`mypy` for static type checking, and `pycodestyle` for PEP 8
conformance.

TODO: Add more type hinting, particularly in `node_to_policy.policy`

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Fixed linter errors

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Improve some docstring documentation

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Adding/fixing docstrings in `nodl_to_policy.policy`

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Bump package version

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Make `convert` CLI more user-friendly

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Adding `convert` verb unit tests

TODO:
* Fix `test_fails_invalid_policy` test. Not sure why I cannot
mock the `load_policy` function with a `etree.InvalidDocument`
side-effect.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Adding `nodl_to_policy` command unit test

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Adding CHANGELOG to the project

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Removing CHANGELOG.rst, no Bloom release yet

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Removing extra whitespace in package.xml

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

aprotyas committed:
Adding common permissions to each node (#7)

* Adding default service/topic/action to each node

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Default permissions generation

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Fixed the duplication of common permissions tags

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Correctly installing data files through setuptools

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Make transitive dependency on lxml explicit

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Improved documentation in `nodl_to_policy.policy`

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Reusing `add_permission` for common permissions

Also, fixing namespace and enclave paths to "/" rather than blank.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Update expected test output policy

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Fixing linter errors

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Fixing more linter errors

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Fixing alphabetic ordering in package listings

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Explicitly use of context manager in `get_profile`

This commit avoids a layer of indirection in having to
call the `_get_profile_path`, which opens a context manager
only to return a file path immediately.
Instead, `_get_profile_path` is removed completely and
the `with` statement is open-coded into the `get_profile`
function, which is the only place that context manager
is required.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Removing memoization pattern from _common._profile

The global Python objects for the common permissions was
an attempt at memoization, but with an unusual pattern.
Given there was no reason to actually try to optimize, and
since the `common_[role]_[permission]` functions will be
called very infrequently, that design pattern can be
eliminated entirely, calling `_get_items_by_role` as required.

NB: this design pattern might require further modification
in the future if lifecycle nodes are ever formalized as a
concept into this tool.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Remove unnecessary 'Optional' type hint

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

aprotyas committed:
Added build/usage instructions in project README (#10)

* Added build/run instructions in project README

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Remove trailing fullstops in build instructions

Improves snippet clarity.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Added programmatic use snippet in README

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Fixing `git clone` instructions in README

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* (misc) Improving README wording

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Formatted README to one sentence per line

As per the Contributing guide:
https://docs.ros.org/en/rolling/Contributing/Code-Style-Language-Versions.html#markdown-restructured-text-docblocks

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Changed right arrow representation for legibility

Changed from `&#8594;` to `&rarr;`. Github renders both perfectly fine.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

aprotyas committed:
Removing a layer of indirection in rule type dictionaries (#12)

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

aprotyas committed:
Removing optional output filename from `convert` verb usage (#11)

* Removing optional output filename with a default

This change achieves the following:

* Simplifies the `convert` verb usage.
* Gets rid of some tricky policy XML file handling business.
* Ensures safer behavior of the `convert` verb, in that an
existing file is not overwritten in any circumstance (unless
the user specifically redirects to that effect).

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

* Changed `convert` verb tests to reflect new usage

The unit tests do not reflect an output policy XML file anymore,
which simplifies both the testing logic (no use of `tmp_path_factory`
anymore) and the number of tests required.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Fixing linter errors

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Merging unrelated histories between aprotyas/nodl_to_policy and aprotyas/nodl_to_policy_osrf

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

aprotyas commented:
Ping @clalancette, @wjwwood

clalancette commented on nodl_to_policy/_common/_profile.py:
```diff
@@ -0,0 +1,61 @@
+
+
+def _get_profile(filename: str) -> etree.ElementTree:
+    # Parses the `node.xml` or the `lifecycle_node.xml` file
+    # The latter filename is WIP, check TODO above in source
```
I'd just remove this comment, since this particular function doesn't "know" anything about `node.xml` or `lifecycle_node.xml`; it is passed as a parameter.

aprotyas replied:
Fixed in 3537e15.

clalancette commented on nodl_to_policy/_common/_profile.py:
```diff
@@ -0,0 +1,61 @@
+    import importlib_resources  # type: ignore
+
+
+# For now, only use `node.xml`, since NoDL does not describe Lifecycle nodes
```
Maybe leave a TODO comment here?

aprotyas replied:
Fixed in 3537e15.

clalancette commented on nodl_to_policy/_common/_profile.py:
```diff
@@ -0,0 +1,61 @@
+    return profile
+
+
+def _get_items_by_role(item_type: str, role: str) -> List:
```
What happens in this function if `item_type` or `role` are the empty string?

aprotyas replied:
Good point. If `item_type` is an empty string, it returns an empty list for the common profile - although really it's looking for an element with an empty tag, which is not what we want to happen. The latter throws an error because of a predicate error - again, not something we want to happen.

As such, I think it's best to return an empty list whenever either of the two arguments are empty. Something like:

```suggestion
def _get_items_by_role(item_type: str, role: str) -> List:
    if not item_type or not role:
        return []
```

aprotyas replied:
Addressed in c11f6e7.

clalancette commented on nodl_to_policy/_common/lifecycle_node.xml:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Since we aren't including functionality for the lifecycle_node yet, should we just remove this file completely?

aprotyas replied:
I believe we should. I wasn't sure about the timeline of NoDL design discussion, so I left it in for now.

clalancette replied:
Here's my suggestion.  We take the lifecycle_node.xml out of this PR (and the Enum).  Then we merge this PR.  Then we open another PR adding the lifecycle support back as a draft.  That way the code exists, but we don't have essentially "dead" code hanging around.

What do you think?

aprotyas replied:
Removed in aa269e5.

aprotyas replied:
> Here's my suggestion. We take the lifecycle_node.xml out of this PR (and the Enum). Then we merge this PR. Then we open another PR adding the lifecycle support back as a draft. That way the code exists, but we don't have essentially "dead" code hanging around.

Yes, I hadn't thought of this option. It's better than leaving the Enum around like in aa269e5. I will remove all of it from this PR and open a draft later then.

aprotyas replied:
Removed both lifecycle_node.xml and the Enum in 531adac. Will open a draft PR related to this topic.

clalancette commented on nodl_to_policy/_verb/_convert.py:
```diff
@@ -0,0 +1,62 @@
+            print(e, file=sys.stderr)
+            return 1
+
+        policy = convert_to_policy(nodl_description)
+        print_policy(policy)
```
Nit, but you can do:

```suggestion
        print_policy(convert_to_policy(nodl_description))
```

clalancette reviewed (changes requested):
I've left a few minor comments of things to think about.  Overall, though, a very nice job!

Commit:
Made policy conversion and printing calls more concise

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Add TODO for lifecycle nodes in `_common._profile`

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Termination of `_get_items_by_role` on empty query

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Remove lifecycle node profile XML and types Enum

These will be drafted (and incorporated) later once the NoDL
design change discussion is formalized.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>
